Change Request: eslint-disable-next-line rule name should not be literal

### ESLint version

v8.49.0

### What problem do you want to solve?

The `eslint-disable-next-line` magic comment feature treats the rule name as literal characters and doesn't handle quoted strings, whereas the `eslint` magic comment processes quoted strings.

This creates an discrepancy between the two commands.

To provide a user story:

I wanted to disable a rule for a single line of code, and searching initially led me to the [**Configure Rules**](https://web.archive.org/web/20230922165926/https://eslint.org/docs/latest/use/configure/rules#rules-from-plugins) page where there was the following quote:

> You can also use this format with configuration comments, such as:
>
> ```
> /* eslint "plugin1/rule1": "error" */
> ```

This led me to believe that rules for third-party plugins (i.e. namespaced packages) required quotation marks, so that's what I did. Then I noticed the rule was applied globally..

That's when I found out about the `eslint-disable-next-line` comment, and I tried changing the comment from:

```js
/* eslint "@typescript-eslint/no-explicit-any": "off" */
```

to:

```js
/* eslint-disable-next-line "@typescript-eslint/no-explicit-any" */
```

and I was met by the following error:

```
/…/src/index.ts
…
  59:3   error  Definition for rule '"@typescript-eslint/no-explicit-any"' was not found            "@typescript-eslint/no-explicit-any"
```

### What do you think is the correct solution?

The `eslint-disable-next-line` magic comment should add support for stringy syntax, such that the following comments all have the same effect:

* `/* eslint-disable-next-line @typescript-eslint/no-explicit-any */`
* `/* eslint-disable-next-line "@typescript-eslint/no-explicit-any" */`
* `/* eslint-disable-next-line '@typescript-eslint/no-explicit-any' */`

### Participation

- [ ] I am willing to submit a pull request for this change.

### Additional comments

_No response_

--- a/lib/linter/apply-disable-directives.js
+++ b/lib/linter/apply-disable-directives.js
@@ -87,7 +87,7 @@ function createIndividualDirectivesRemoval(directives, commentToken) {
     return directives.map(directive => {
         const { ruleId } = directive;
 
-        const regex = new RegExp(String.raw`(?:^|\s*,\s*)${escapeRegExp(ruleId)}(?:\s*,\s*|$)`, "u");
+        const regex = new RegExp(String.raw`(?:^|\s*,\s*)(?<quote>['"]?)${escapeRegExp(ruleId)}\k<quote>(?:\s*,\s*|$)`, "u");
         const match = regex.exec(listText);
         const matchedText = match[0];
         const matchStartOffset = listStartOffset + match.index;

--- a/lib/linter/config-comment-parser.js
+++ b/lib/linter/config-comment-parser.js
@@ -139,7 +139,7 @@ module.exports = class ConfigCommentParser {
         const items = {};
 
         string.split(",").forEach(name => {
-            const trimmedName = name.replace(/^\s*(?<quote>['"]?)(?<ruleId>\S*)\k<quote>\s*$/u, "$<ruleId>");
+            const trimmedName = name.trim().replace(/^(?<quote>['"]?)(?<ruleId>.*)\k<quote>$/us, "$<ruleId>");
 
             if (trimmedName) {
                 items[trimmedName] = true;

--- a/lib/linter/config-comment-parser.js
+++ b/lib/linter/config-comment-parser.js
@@ -139,7 +139,7 @@ module.exports = class ConfigCommentParser {
         const items = {};
 
         string.split(",").forEach(name => {
-            const trimmedName = name.trim();
+            const trimmedName = name.replace(/^\s*(?<quote>['"]?)(?<ruleId>\S*)\k<quote>\s*$/u, "$<ruleId>");
 
             if (trimmedName) {
                 items[trimmedName] = true;

--- a/tests/lib/linter/config-comment-parser.js
+++ b/tests/lib/linter/config-comment-parser.js
@@ -236,6 +236,16 @@ describe("ConfigCommentParser", () => {
                 "'c\"": true // This result is correct because used mismatched quotes.
             });
         });
+        it("should parse list config with spaced items", () => {
+            const code = " a b , 'c d' , \"e f\" ";
+            const result = commentParser.parseListConfig(code);
+
+            assert.deepStrictEqual(result, {
+                "a b": true,
+                "c d": true,
+                "e f": true
+            });
+        });
     });
 
 });

--- a/tests/lib/linter/config-comment-parser.js
+++ b/tests/lib/linter/config-comment-parser.js
@@ -224,6 +224,18 @@ describe("ConfigCommentParser", () => {
                 b: true
             });
         });
+
+        it("should parse list config with literal items", () => {
+            const code = "'a', \"b\", 'c\", \"d'";
+            const result = commentParser.parseListConfig(code);
+
+            assert.deepStrictEqual(result, {
+                a: true,
+                b: true,
+                "\"d'": true, // This result is correct because used mismatched quotes.
+                "'c\"": true // This result is correct because used mismatched quotes.
+            });
+        });
     });
 
 });

--- a/tests/lib/linter/config-comment-parser.js
+++ b/tests/lib/linter/config-comment-parser.js
@@ -225,7 +225,7 @@ describe("ConfigCommentParser", () => {
             });
         });
 
-        it("should parse list config with literal items", () => {
+        it("should parse list config with quoted items", () => {
             const code = "'a', \"b\", 'c\", \"d'";
             const result = commentParser.parseListConfig(code);
 

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -15473,15 +15473,15 @@ var a = "test2";
                         });
 
                         // Test for literals
-                        for (const testcaseForliteral of [
+                        for (const testcaseForLiteral of [
                             { code: code.replace(/(test\/[\w-]+)/gu, '"$1"'), output: output.replace(/(test\/[\w-]+)/gu, '"$1"') },
                             { code: code.replace(/(test\/[\w-]+)/gu, "'$1'"), output: output.replace(/(test\/[\w-]+)/gu, "'$1'") }
                         ]) {
                             // eslint-disable-next-line no-loop-func -- `linter` is getting updated in beforeEach()
-                            it(testcaseForliteral.code, () => {
+                            it(testcaseForLiteral.code, () => {
                                 assert.strictEqual(
-                                    linter.verifyAndFix(testcaseForliteral.code, config).output,
-                                    testcaseForliteral.output
+                                    linter.verifyAndFix(testcaseForLiteral.code, config).output,
+                                    testcaseForLiteral.output
                                 );
                             });
                         }

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -4949,8 +4949,8 @@ var a = "test2";
 
                 // Test for quoted rule names
                 for (const testcaseForLiteral of [
-                    { code: code.replace(/(test\/[\w-]+)/gu, '"$1"'), output: output.replace(/(test\/[\w-]+)/gu, '"$1"') },
-                    { code: code.replace(/(test\/[\w-]+)/gu, "'$1'"), output: output.replace(/(test\/[\w-]+)/gu, "'$1'") }
+                    { code: code.replace(/((?:un)?used[\w-]*)/gu, '"$1"'), output: output.replace(/((?:un)?used[\w-]*)/gu, '"$1"') },
+                    { code: code.replace(/((?:un)?used[\w-]*)/gu, "'$1'"), output: output.replace(/((?:un)?used[\w-]*)/gu, "'$1'") }
                 ]) {
                     // eslint-disable-next-line no-loop-func -- `linter` is getting updated in beforeEach()
                     it(testcaseForLiteral.code, () => {

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -13060,6 +13060,60 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages[1].line, 3);
                 });
 
+                it("should report a violation with literals", () => {
+                    const code = [
+                        "/*eslint-disable 'no-alert' */",
+                        "alert('test');",
+                        "console.log('test');", // here
+                        "/*eslint-enable */",
+                        "/*eslint-disable \"no-console\" */",
+                        "alert('test');", // here
+                        "console.log('test');"
+                    ].join("\n");
+                    const config = { rules: { "no-alert": 1, "no-console": 1 } };
+
+                    const messages = linter.verify(code, config, filename);
+                    const suppressedMessages = linter.getSuppressedMessages();
+
+                    assert.strictEqual(messages.length, 2);
+                    assert.strictEqual(messages[0].ruleId, "no-console");
+                    assert.strictEqual(messages[1].ruleId, "no-alert");
+
+                    assert.strictEqual(suppressedMessages.length, 2);
+                    assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+                    assert.strictEqual(suppressedMessages[1].ruleId, "no-console");
+                });
+
+                it("should report a violation with literals", () => {
+                    const code = [
+                        "/*eslint-disable no-alert, no-console */",
+                        "alert('test');",
+                        "console.log('test');",
+                        "/*eslint-enable 'no-alert'*/",
+                        "alert('test');", // here
+                        "console.log('test');",
+                        "/*eslint-enable \"no-console\"*/",
+                        "console.log('test');" // here
+                    ].join("\n");
+                    const config = { rules: { "no-alert": 1, "no-console": 1 } };
+
+                    const messages = linter.verify(code, config, filename);
+                    const suppressedMessages = linter.getSuppressedMessages();
+
+                    assert.strictEqual(messages.length, 2);
+                    assert.strictEqual(messages[0].ruleId, "no-alert");
+                    assert.strictEqual(messages[0].line, 5);
+                    assert.strictEqual(messages[1].ruleId, "no-console");
+                    assert.strictEqual(messages[1].line, 8);
+
+                    assert.strictEqual(suppressedMessages.length, 3);
+                    assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+                    assert.strictEqual(suppressedMessages[0].line, 2);
+                    assert.strictEqual(suppressedMessages[1].ruleId, "no-console");
+                    assert.strictEqual(suppressedMessages[1].line, 3);
+                    assert.strictEqual(suppressedMessages[2].ruleId, "no-console");
+                    assert.strictEqual(suppressedMessages[2].line, 6);
+                });
             });
 
             describe("/*eslint-disable-line*/", () => {
@@ -13293,6 +13347,32 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages.length, 5);
                 });
 
+                it("should report a violation with literals", () => {
+                    const code = [
+                        "alert('test'); // eslint-disable-line 'no-alert'",
+                        "console.log('test');", // here
+                        "alert('test'); // eslint-disable-line \"no-alert\""
+                    ].join("\n");
+                    const config = {
+                        rules: {
+                            "no-alert": 1,
+                            "no-console": 1
+                        }
+                    };
+
+                    const messages = linter.verify(code, config, filename);
+                    const suppressedMessages = linter.getSuppressedMessages();
+
+                    assert.strictEqual(messages.length, 1);
+                    assert.strictEqual(messages[0].ruleId, "no-console");
+                    assert.strictEqual(messages[0].line, 2);
+
+                    assert.strictEqual(suppressedMessages.length, 2);
+                    assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+                    assert.strictEqual(suppressedMessages[0].line, 1);
+                    assert.strictEqual(suppressedMessages[1].ruleId, "no-alert");
+                    assert.strictEqual(suppressedMessages[1].line, 3);
+                });
             });
 
             describe("/*eslint-disable-next-line*/", () => {
@@ -13688,6 +13768,31 @@ var a = "test2";
                     assert.strictEqual(messages[1].ruleId, "no-console");
 
                     assert.strictEqual(suppressedMessages.length, 0);
+                });
+
+                it("should ignore violation of specified rule on next line with literals", () => {
+                    const code = [
+                        "// eslint-disable-next-line 'no-alert'",
+                        "alert('test');",
+                        "// eslint-disable-next-line \"no-alert\"",
+                        "alert('test');",
+                        "console.log('test');"
+                    ].join("\n");
+                    const config = {
+                        rules: {
+                            "no-alert": 1,
+                            "no-console": 1
+                        }
+                    };
+                    const messages = linter.verify(code, config, filename);
+                    const suppressedMessages = linter.getSuppressedMessages();
+
+                    assert.strictEqual(messages.length, 1);
+                    assert.strictEqual(messages[0].ruleId, "no-console");
+
+                    assert.strictEqual(suppressedMessages.length, 2);
+                    assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+                    assert.strictEqual(suppressedMessages[1].ruleId, "no-alert");
                 });
             });
 
@@ -15366,6 +15471,20 @@ var a = "test2";
                                 output
                             );
                         });
+
+                        // Test for literals
+                        for (const testcaseForliteral of [
+                            { code: code.replace(/(test\/[\w-]+)/gu, '"$1"'), output: output.replace(/(test\/[\w-]+)/gu, '"$1"') },
+                            { code: code.replace(/(test\/[\w-]+)/gu, "'$1'"), output: output.replace(/(test\/[\w-]+)/gu, "'$1'") }
+                        ]) {
+                            // eslint-disable-next-line no-loop-func -- `linter` is getting updated in beforeEach()
+                            it(testcaseForliteral.code, () => {
+                                assert.strictEqual(
+                                    linter.verifyAndFix(testcaseForliteral.code, config).output,
+                                    testcaseForliteral.output
+                                );
+                            });
+                        }
                     }
                 });
             });

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -2604,6 +2604,33 @@ describe("Linter", () => {
                 assert.strictEqual(suppressedMessages.length, 2);
                 assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
             });
+
+            it("should report a violation with quoted rule names in eslint-disable-line", () => {
+                const code = [
+                    "alert('test'); // eslint-disable-line 'no-alert'",
+                    "console.log('test');", // here
+                    "alert('test'); // eslint-disable-line \"no-alert\""
+                ].join("\n");
+                const config = {
+                    rules: {
+                        "no-alert": 1,
+                        "no-console": 1
+                    }
+                };
+
+                const messages = linter.verify(code, config, filename);
+                const suppressedMessages = linter.getSuppressedMessages();
+
+                assert.strictEqual(messages.length, 1);
+                assert.strictEqual(messages[0].ruleId, "no-console");
+                assert.strictEqual(messages[0].line, 2);
+
+                assert.strictEqual(suppressedMessages.length, 2);
+                assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+                assert.strictEqual(suppressedMessages[0].line, 1);
+                assert.strictEqual(suppressedMessages[1].ruleId, "no-alert");
+                assert.strictEqual(suppressedMessages[1].line, 3);
+            });
         });
 
         describe("eslint-disable-next-line", () => {
@@ -2957,6 +2984,31 @@ describe("Linter", () => {
 
                 assert.strictEqual(suppressedMessages.length, 0);
             });
+
+            it("should ignore violation of specified rule on next line with quoted rule names", () => {
+                const code = [
+                    "// eslint-disable-next-line 'no-alert'",
+                    "alert('test');",
+                    "// eslint-disable-next-line \"no-alert\"",
+                    "alert('test');",
+                    "console.log('test');"
+                ].join("\n");
+                const config = {
+                    rules: {
+                        "no-alert": 1,
+                        "no-console": 1
+                    }
+                };
+                const messages = linter.verify(code, config, filename);
+                const suppressedMessages = linter.getSuppressedMessages();
+
+                assert.strictEqual(messages.length, 1);
+                assert.strictEqual(messages[0].ruleId, "no-console");
+
+                assert.strictEqual(suppressedMessages.length, 2);
+                assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+                assert.strictEqual(suppressedMessages[1].ruleId, "no-alert");
+            });
         });
     });
 
@@ -3224,6 +3276,61 @@ describe("Linter", () => {
             assert.strictEqual(messages[1].line, 8);
             assert.strictEqual(messages[2].ruleId, "no-console");
             assert.strictEqual(messages[2].line, 9);
+
+            assert.strictEqual(suppressedMessages.length, 3);
+            assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+            assert.strictEqual(suppressedMessages[0].line, 2);
+            assert.strictEqual(suppressedMessages[1].ruleId, "no-console");
+            assert.strictEqual(suppressedMessages[1].line, 3);
+            assert.strictEqual(suppressedMessages[2].ruleId, "no-console");
+            assert.strictEqual(suppressedMessages[2].line, 6);
+        });
+
+        it("should report a violation with quoted rule names in eslint-disable", () => {
+            const code = [
+                "/*eslint-disable 'no-alert' */",
+                "alert('test');",
+                "console.log('test');", // here
+                "/*eslint-enable */",
+                "/*eslint-disable \"no-console\" */",
+                "alert('test');", // here
+                "console.log('test');"
+            ].join("\n");
+            const config = { rules: { "no-alert": 1, "no-console": 1 } };
+
+            const messages = linter.verify(code, config, filename);
+            const suppressedMessages = linter.getSuppressedMessages();
+
+            assert.strictEqual(messages.length, 2);
+            assert.strictEqual(messages[0].ruleId, "no-console");
+            assert.strictEqual(messages[1].ruleId, "no-alert");
+
+            assert.strictEqual(suppressedMessages.length, 2);
+            assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
+            assert.strictEqual(suppressedMessages[1].ruleId, "no-console");
+        });
+
+        it("should report a violation with quoted rule names in eslint-enable", () => {
+            const code = [
+                "/*eslint-disable no-alert, no-console */",
+                "alert('test');",
+                "console.log('test');",
+                "/*eslint-enable 'no-alert'*/",
+                "alert('test');", // here
+                "console.log('test');",
+                "/*eslint-enable \"no-console\"*/",
+                "console.log('test');" // here
+            ].join("\n");
+            const config = { rules: { "no-alert": 1, "no-console": 1 } };
+
+            const messages = linter.verify(code, config, filename);
+            const suppressedMessages = linter.getSuppressedMessages();
+
+            assert.strictEqual(messages.length, 2);
+            assert.strictEqual(messages[0].ruleId, "no-alert");
+            assert.strictEqual(messages[0].line, 5);
+            assert.strictEqual(messages[1].ruleId, "no-console");
+            assert.strictEqual(messages[1].line, 8);
 
             assert.strictEqual(suppressedMessages.length, 3);
             assert.strictEqual(suppressedMessages[0].ruleId, "no-alert");
@@ -4813,6 +4920,20 @@ var a = "test2";
                         output
                     );
                 });
+
+                // Test for quoted rule names
+                for (const testcaseForLiteral of [
+                    { code: code.replace(/(test\/[\w-]+)/gu, '"$1"'), output: output.replace(/(test\/[\w-]+)/gu, '"$1"') },
+                    { code: code.replace(/(test\/[\w-]+)/gu, "'$1'"), output: output.replace(/(test\/[\w-]+)/gu, "'$1'") }
+                ]) {
+                    // eslint-disable-next-line no-loop-func -- `linter` is getting updated in beforeEach()
+                    it(testcaseForLiteral.code, () => {
+                        assert.strictEqual(
+                            linter.verifyAndFix(testcaseForLiteral.code, config).output,
+                            testcaseForLiteral.output
+                        );
+                    });
+                }
             }
         });
     });
@@ -13060,7 +13181,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages[1].line, 3);
                 });
 
-                it("should report a violation with literals in eslint-disable", () => {
+                it("should report a violation with quoted rule names in eslint-disable", () => {
                     const code = [
                         "/*eslint-disable 'no-alert' */",
                         "alert('test');",
@@ -13084,7 +13205,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages[1].ruleId, "no-console");
                 });
 
-                it("should report a violation with literals in eslint-enable", () => {
+                it("should report a violation with quoted rule names in eslint-enable", () => {
                     const code = [
                         "/*eslint-disable no-alert, no-console */",
                         "alert('test');",
@@ -13347,7 +13468,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages.length, 5);
                 });
 
-                it("should report a violation with literals in eslint-disable-line", () => {
+                it("should report a violation with quoted rule names in eslint-disable-line", () => {
                     const code = [
                         "alert('test'); // eslint-disable-line 'no-alert'",
                         "console.log('test');", // here
@@ -13770,7 +13891,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages.length, 0);
                 });
 
-                it("should ignore violation of specified rule on next line with literals", () => {
+                it("should ignore violation of specified rule on next line with quoted rule names", () => {
                     const code = [
                         "// eslint-disable-next-line 'no-alert'",
                         "alert('test');",
@@ -15472,7 +15593,7 @@ var a = "test2";
                             );
                         });
 
-                        // Test for literals
+                        // Test for quoted rule names
                         for (const testcaseForLiteral of [
                             { code: code.replace(/(test\/[\w-]+)/gu, '"$1"'), output: output.replace(/(test\/[\w-]+)/gu, '"$1"') },
                             { code: code.replace(/(test\/[\w-]+)/gu, "'$1'"), output: output.replace(/(test\/[\w-]+)/gu, "'$1'") }

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -13060,7 +13060,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages[1].line, 3);
                 });
 
-                it("should report a violation with literals", () => {
+                it("should report a violation with literals in eslint-disable", () => {
                     const code = [
                         "/*eslint-disable 'no-alert' */",
                         "alert('test');",
@@ -13084,7 +13084,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages[1].ruleId, "no-console");
                 });
 
-                it("should report a violation with literals", () => {
+                it("should report a violation with literals in eslint-enable", () => {
                     const code = [
                         "/*eslint-disable no-alert, no-console */",
                         "alert('test');",
@@ -13347,7 +13347,7 @@ var a = "test2";
                     assert.strictEqual(suppressedMessages.length, 5);
                 });
 
-                it("should report a violation with literals", () => {
+                it("should report a violation with literals in eslint-disable-line", () => {
                     const code = [
                         "alert('test'); // eslint-disable-line 'no-alert'",
                         "console.log('test');", // here

--- a/tests/lib/linter/linter.js
+++ b/tests/lib/linter/linter.js
@@ -1467,6 +1467,32 @@ describe("Linter", () => {
             linter.verify(code, config);
             assert(spy && spy.calledOnce);
         });
+
+        it("variables should be available in global scope with quoted items", () => {
+            const code = `/*${ESLINT_ENV} 'node'*/ function f() {} /*${ESLINT_ENV} "browser", "mocha"*/`;
+            const config = { rules: { checker: "error" } };
+            let spy;
+
+            linter.defineRule("checker", {
+                create(context) {
+                    spy = sinon.spy(() => {
+                        const scope = context.getScope(),
+                            exports = getVariable(scope, "exports"),
+                            window = getVariable(scope, "window"),
+                            it = getVariable(scope, "it");
+
+                        assert.strictEqual(exports.writeable, true);
+                        assert.strictEqual(window.writeable, false);
+                        assert.strictEqual(it.writeable, false);
+                    });
+
+                    return { Program: spy };
+                }
+            });
+
+            linter.verify(code, config);
+            assert(spy && spy.calledOnce);
+        });
     });
 
     describe("when evaluating code containing /*eslint-env */ block with sloppy whitespace", () => {

